app/scripts/controllers/network/fake-provider.ts

```diff
@@ -0,0 +1,224 @@
+import { inspect, isDeepStrictEqual } from 'util';
+import {
+  JsonRpcEngine,
+  JsonRpcSuccess,
+  JsonRpcFailure,
+  JsonRpcRequest,
+  JsonRpcResponse,
+} from 'json-rpc-engine';
+import { SafeEventEmitterProvider } from '@metamask/eth-json-rpc-provider/dist/safe-event-emitter-provider';
+
+// Store this in case it gets stubbed later
+const originalSetTimeout = global.setTimeout;
+
+/**
+ * An object that allows specifying the behavior of a specific invocation of
+ * `sendAsync`. The `method` always identifies the stub, but the behavior
+ * may be specified multiple ways: `sendAsync` can either return a promise or
+ * throw an error, and if it returns a promise, that promise can either be
+ * resolved with a response object or reject with an error.
+ *
+ * @property request - Looks for a request matching these specifications.
+ * @property request.method - The RPC method to which this stub will be matched.
+ * @property request.params - The params to which this stub will be matched.
+ * @property response - Instructs `sendAsync` to return a promise that resolves
+ * with a response object.
+ * @property response.result - Specifies a successful response, with this as the
+ * `result`.
+ * @property response.error - Specifies an error response, with this as the
+ * `error`.
+ * @property error - Instructs `sendAsync` to return a promise that rejects with
+ * this error.
+ * @property implementation - Allows overriding `sendAsync` entirely. Useful if
+ * you want it to throw an error.
+ * @property delay - The amount of time that will pass after the callback is
+ * called with the response.
+ * @property discardAfterMatching - Usually after the stub matches a request, it
+ * is discarded, but setting this to true prevents that from happening. True by
+ * default.
+ */
+export type FakeProviderStub<Params, Response> = {
```

Comment:
It doesn't seem like these generics are serving any purpose. I see that they're carried through to the `FakeProvider` class `send` and `sendAsync` methods, but those methods are never called from a context that would be aware of the stub types used. Effectively they may as well be `any` here since they're set as `any` in the fake provider engine, making these less type safe than encoding the constraints in this type instead of generics.

Reply:
Yeah... you're right. I've removed them here: [9bf8177](https://github.com/MetaMask/metamask-extension/pull/18628/commits/9bf81775e16a2109197ac7c35a67a1c515fc27ae)